kvstreamer: fix cross-range ScanRequests with SingleRowLookup hint

yuzefovich opened the pull request:
This commit fixes a long-standing issue in the streamer when SingleRowLookup hint is set and there are some cross-range ScanRequests. The SingleRowLookup hint indicates that each request can result in at most one SQL row, but that was incorrectly understood as ScanRequests touching exactly one range - it is possible to have a multi-range ScanRequest that only fetches a single SQL row. As a result of this incorrect usage of the SingleRowLookup hint, the streamer could either get stuck (because it would actually receive more responses than were expected, and in `resultsBuffer.get` implementations we say `allComplete` is `true` with exact equality) or could return incorrect results (when one of the truncated ScanRequests came back empty before the non-empty one, but we didn't wait for the latter).

The bug is now fixed by removing optimizations that were in-place when this hint is specified. In particular, it means that now, whenever at least one ScanRequest is enqueued, we will always allocate
- `Streamer.mu.numRangesPerScanRequest []int32` slice
- `subRequestIdx []int32` slices in the InOrder mode.

This would increase the fraction of the streamer's budget dedicated to overhead, but it should be relatively minor in the grand scheme of things.

I decided to not remove the hint altogether since we could introduce some optimizations using it in the future.

Fixes: #101823.
Fixes: #108412.
Fixes: #112490.

Release note (bug fix): Previously, CockroachDB could incorrectly evaluate queries with lookup joins when equality columns are key in some rare cases when looking up into multiple ranges. Sometimes this would manifest as a stuck query, sometimes it could result in incorrect output. The bug was introduced in 22.2 and is now fixed.

yuzefovich committed:
kvstreamer: fix cross-range ScanRequests with SingleRowLookup hint

This commit fixes a long-standing issue in the streamer when
SingleRowLookup hint is set and there are some cross-range ScanRequests.
The SingleRowLookup hint indicates that each request can result in at
most one SQL row, but that was incorrectly understood as ScanRequests
touching exactly one range - it is possible to have a multi-range
ScanRequest that only fetches a single SQL row. As a result of this
incorrect usage of the SingleRowLookup hint, the streamer could either
get stuck (because it would actually receive more responses than were
expected, and in `resultsBuffer.get` implementations we say `allComplete`
is `true` with exact equality) or could return incorrect results (when
one of the truncated ScanRequests came back empty before the non-empty
one, but we didn't wait for the latter).

The bug is now fixed by removing optimizations that were in-place when
this hint is specified. In particular, it means that now, whenever at
least one ScanRequest is enqueued, we will always allocate
- `Streamer.mu.numRangesPerScanRequest []int32` slice
- `subRequestIdx []int32` slices in the InOrder mode.

This would increase the fraction of the streamer's budget dedicated to
overhead, but it should be relatively minor in the grand scheme of
things.

I decided to not remove the hint altogether since we could introduce
some optimizations using it in the future.

Release note (bug fix): Previously, CockroachDB could incorrectly
evaluate queries with lookup joins when equality columns are key in some
rare cases when looking up into multiple ranges. Sometimes this would
manifest as a stuck query, sometimes it could result in incorrect
output. The bug was introduced in 22.2 and is now fixed.

yuzefovich committed:
kvstreamer: add a couple of assertions

This commit adds a couple of assertions that fire whenever the
resultsBuffer receives more complete responses than it expected. These
assertions would have prevented deadlocks fixed in the previous commit.

Also it adds some more details to existing assertions.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/115569)
<!-- Reviewable:end -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-:-Nkr1_Tw0yoEOd9EN6FI:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/kv/kvclient/kvstreamer/streamer.go` line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-Nkr-n1AC8eJt2EaJoD0:-Nkr-n1AC8eJt2EaJoD1:bz375af) ([raw file](https://github.com/cockroachdb/cockroach/blob/9103fe6f1ad10f542d474abcc67657bb37d387fc/pkg/kv/kvclient/kvstreamer/streamer.go#L627)):*
> ```Go
> 				// TODO(yuzefovich): we could avoid using / allocating
> 				// numRangesPerScanRequest if allRequestsAreWithinSingleRange is
> 				// true.
> ```

Could we go even further and have `Intersect` check whether each span was actually truncated? That way, we could also support the case when the batch spans multiple ranges, but individual requests don't. That seems like it would be common for large tables.
___
*[`pkg/sql/logictest/testdata/logic_test/lookup_join` line 1573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-Nkr0mNH1va7LycpAckp:-Nkr0mNH1va7LycpAckq:bcx2f52) ([raw file](https://github.com/cockroachdb/cockroach/blob/9103fe6f1ad10f542d474abcc67657bb37d387fc/pkg/sql/logictest/testdata/logic_test/lookup_join#L1573)):*
> ```
> CREATE TABLE r_101823 (k INT PRIMARY KEY, u INT, v INT, w BOOL, UNIQUE INDEX (u), UNIQUE INDEX (u, v, w));
> INSERT INTO r_101823 VALUES (1, 1, 1, false);
> ALTER INDEX r_101823_u_v_w_key SPLIT AT VALUES (1, 1, false), (1, 1, true);
> ```

Great find with this test! It's confusing that we allow this, but I guess disallowing it could be even more confusing for users...


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I think it's likely that this fixes two sentry issues that we've seen, so I'm optimistically closing them. I also made some minor adjustments to the second commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-:-Nkr_6VS3bZMN1c-Jfc4:b-k6w2x2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/kv/kvclient/kvstreamer/streamer.go` line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-Nkr-n1AC8eJt2EaJoD0:-NkrYLIE9Wbkw_TKy0jF:bvqtcxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9103fe6f1ad10f542d474abcc67657bb37d387fc/pkg/kv/kvclient/kvstreamer/streamer.go#L627)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could we go even further and have `Intersect` check whether each span was actually truncated? That way, we could also support the case when the batch spans multiple ranges, but individual requests don't. That seems like it would be common for large tables.
</blockquote></details>

Yeah, good point, adjusted the TODO.
___
*[`pkg/sql/logictest/testdata/logic_test/lookup_join` line 1573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-Nkr0mNH1va7LycpAckp:-NkrYhKKDEUhWN0MvuBd:bdxefro) ([raw file](https://github.com/cockroachdb/cockroach/blob/9103fe6f1ad10f542d474abcc67657bb37d387fc/pkg/sql/logictest/testdata/logic_test/lookup_join#L1573)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Great find with this test! It's confusing that we allow this, but I guess disallowing it could be even more confusing for users...
</blockquote></details>

Can you say a bit more about what's confusing? The fact that a ScanRequest is guaranteed to have at most one SQL row but is not guaranteed to be within a single range?


<!-- Sent from Reviewable.io -->

yuzefovich committed:
kvstreamer: remove no longer needed special handling of ctx cancel

This code was added in the hopes of reproducing the issue fixed in the
previous commit. The check for context cancellation is kept in
`resultsBuffer.wait` implementations just to be safe - it's likely no
longer needed, but it's better to have it in case other deadlock
scenarios exist.

Release note: None

yuzefovich reviewed (commented):
Also added another small third commit to clean up some things. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-:-NkrnZ_PBLLv1wUx4h_4:b7gfwuj)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r3, 2 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-:-NkrpEgx0FTelvThly7Y:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/logictest/testdata/logic_test/lookup_join` line 1573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-Nkr0mNH1va7LycpAckp:-Nkrms-s0qPAVBBZZQWt:b-8gm7qk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9103fe6f1ad10f542d474abcc67657bb37d387fc/pkg/sql/logictest/testdata/logic_test/lookup_join#L1573)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Can you say a bit more about what's confusing? The fact that a ScanRequest is guaranteed to have at most one SQL row but is not guaranteed to be within a single range?
</blockquote></details>

Yes, it feels odd that a range can be split on a key that isn't allowed to exist because of a unique constraint. I wonder if there are any other unintuitive range-splitting points similar to this or between column families. That's not a problem with this PR, though.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the quick reviews!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-:-Nkrq_df9cCmMgYwS7JD:bbj06re)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/logictest/testdata/logic_test/lookup_join` line 1573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115569#-Nkr0mNH1va7LycpAckp:-Nkrpmpv0j1shXQ7CI5X:b35px1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9103fe6f1ad10f542d474abcc67657bb37d387fc/pkg/sql/logictest/testdata/logic_test/lookup_join#L1573)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Yes, it feels odd that a range can be split on a key that isn't allowed to exist because of a unique constraint. I wonder if there are any other unintuitive range-splitting points similar to this or between column families. That's not a problem with this PR, though.
</blockquote></details>

Hm, I agree that it's unfortunate, but I doubt we'd be able to do anything about it. At the KV level, splits like in this reproduction are perfectly valid since they don't break the only invariant SQL is asking for (that no SQL row is split across ranges).

For context, I added the UNIQUE constraint on `u` to get "equality columns are key" property for the query, perhaps there are other ways where we can get this property without such an artificial example.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
>For context, I added the UNIQUE constraint on u to get "equality columns are key" property for the query, perhaps there are other ways where we can get this property without such an artificial example.

A more interesting case could have the unique column other than the leading one, like with indexes on `(a)` and `(b, a, c)`. The only other case I can think of that would cause "equality columns are key" would be if the table was guaranteed to only have a single row (e.g. by a check constraint). That seems like an edge case, but there is an internal table that uses that pattern.

I think the existing test is good though, just throwing out ideas.
